### PR TITLE
Add Istio sidecar enabled tests and refactor code

### PR DESCRIPTION
**Pull request by cnvergence**
**Description**

Changes proposed in this pull request:

- Add component test checking lack of Istio sidecars in pod deployed in namespace labeled with istio-injection=enabled
- Merge Istio sidecar injection tests as one feature.
- Clean up the code by splitting istio_installed_test.go to main_test.go
- Add documentation README.md

### CODE REVIEW

File: tests/components/istio/features/sidecar_injection.feature [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  Scenario: Namespace with istio-injection=enabled label does contain pods with istio sidecar
+    Given Istio component is installed
+    When "sidecar-enable" namespace exists
+    And "sidecar-enable" namespace is labeled with "istio-injection" "enabled"
+    And Httpbin deployment is created in "sidecar-enable" namespace
+    And Httpbin deployment is deployed and ready in "sidecar-enable" namespace
+    Then there should be any pods with istio sidecar in "sidecar-enable" namespace
+    And "sidecar-enable" namespace is deleted
+
+  Scenario: Kyma-system namespace does not contain pods with sidecar
```
Comment (barchw) [inline tests/components/istio/features/sidecar_injection.feature:21]:
```suggestion
  Scenario: Kyma-system namespace contains pods with sidecar
```

File: tests/components/istio/features/sidecar_injection.feature [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    Given Istio component is installed
+    Then there should be any pods with istio sidecar in "kyma-system" namespace
```
Comment (barchw) [inline tests/components/istio/features/sidecar_injection.feature:23]:
```suggestion
    Then there should be some pods with istio sidecar in "kyma-system" namespace
```

File: tests/components/istio/istio_sidecar_test.go [hunk 3]
```diff
@@ -20,15 +20,58 @@ func InitializeScenarioTargetNamespaceSidecar(ctx *godog.ScenarioContext) {
 		err := installedCase.getIstioPods()
 		return ctx, err
 	})
+
 	ctx.Step(`^Istio component is installed$`, installedCase.istioComponentIsInstalled)
+	ctx.Step(`^there should be no pods with istio sidecar in "([^"]*)" namespace$`, installedCase.targetNamespacePodsShouldNotHaveSidecar)
+	ctx.Step(`^there should be any pods with istio sidecar in "([^"]*)" namespace$`, installedCase.targetNamespacePodsShouldHaveSidecar)
```
Comment (barchw) [inline tests/components/istio/istio_sidecar_test.go:26]:
```suggestion
	ctx.Step(`^there should be some pods with istio sidecar in "([^"]*)" namespace$`, installedCase.targetNamespacePodsShouldHaveSidecar)
```

File: tests/components/istio/istio_sidecar_test.go [hunk 4]
```diff
@@ -20,15 +20,58 @@ func InitializeScenarioTargetNamespaceSidecar(ctx *godog.ScenarioContext) {
...
 	ctx.Step(`^"([^"]*)" namespace exists`, installedCase.createTargetNamespace)
 	ctx.Step(`^"([^"]*)" namespace is labeled with "([^"]*)" "([^"]*)"$`, installedCase.labelTargetNamespace)
-	ctx.Step(`^Httpbin is deployed in "([^"]*)" namespace$`, installedCase.deployHttpBinInTargetNamespace)
+	ctx.Step(`^Httpbin deployment is created in "([^"]*)" namespace$`, installedCase.deployHttpBinInTargetNamespace)
 	ctx.Step(`^Httpbin deployment is deployed and ready in "([^"]*)" namespace$`, installedCase.waitForHttpBinInTargetNamespace)
 	ctx.Step(`^there should be no pods with istio sidecar in "([^"]*)" namespace$`, installedCase.targetNamespacePodsShouldNotHaveSidecar)
+	ctx.Step(`^there should be any pods with istio sidecar in "([^"]*)" namespace$`, installedCase.targetNamespacePodsShouldHaveSidecar)
```
Comment (barchw) [inline tests/components/istio/istio_sidecar_test.go:32]: This is doubled

File: tests/components/istio/istio_installed_test.go [hunk 5]
```diff
@@ -2,156 +2,43 @@ package istio
...
-	if suite.Name != os.Getenv(deployedKymaProfileVar) {
-		t.Skip()
-	}
-	if suite.Run() != 0 {
-		t.Fatal("non-zero status returned, failed to run feature tests")
-	}
-	if os.Getenv(exportResultVar) == "true" {
-		generateHTMLReport()
-	}
+	ctx.Step(`^a running Kyma cluster with "([^"]*)" profile$`, installedCase.aRunningKymaClusterWithProfile)
```
Comment (barchw) [inline tests/components/istio/istio_installed_test.go:20]: Can we make a func from this? Like InitializeScenarioIstioInstalled(ctx)

File: tests/components/istio/istio_installed_test.go [hunk 6]
```diff
@@ -2,156 +2,43 @@ package istio
...
-type istioInstallledCase struct {
-	pilotPods     *corev1.PodList
-	ingressGwPods *corev1.PodList
+func InitializeScenarioProdProfile(ctx *godog.ScenarioContext) {
+	installedCase := istioInstallledCase{}
+	ctx.Before(func(ctx context.Context, sc *godog.Scenario) (context.Context, error) {
+		err := installedCase.getIstioPods()
+		return ctx, err
+	})
+	ctx.Step(`^a running Kyma cluster with "([^"]*)" profile$`, installedCase.aRunningKymaClusterWithProfile)
```
Comment (barchw) [inline tests/components/istio/istio_installed_test.go:35]: Can we make a func from those steps? Like InitializeScenarioIstioInstalled(ctx)

File: tests/components/istio/main_test.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+	if suiteExitCode != 0 {
+		t.Fatal("non-zero status returned, failed to run feature tests")
+	}
+}
+
+func TestIstioInstalledProduction(t *testing.T) {
+	prodOpts := goDogOpts
+	prodOpts.Paths = []string{"features/istio_production.feature", "features/sidecar_injection.feature"}
```
Comment (barchw) [inline tests/components/istio/main_test.go:63]: Maybe we can "automate" that feature fetching? 
```suggestion
	prodOpts.Paths = []string{}
 	filepath.Walk("features", func(path string, info fs.FileInfo, err error) error {
 		prodOpts.Paths = append(prodOpts.Paths, path)
 		return nil
 	})
```

**Review by barchw — CHANGES_REQUESTED** [review]

File: tests/components/istio/README.md [hunk 8]
```diff
@@ -0,0 +1,42 @@
+# Istio component tests
+
+We have chosen to implement component tests with [cucumber/godog](https://github.com/cucumber/godog) library. 
+Please feel free to take a look at their repository to learn more about the framework.
+
+## Requirements
+
+In order to run these tests, you need to have:
+- Kubernetes installed and kubeconfig configured to point on this cluster
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:9]:
```suggestion
- Kubernetes installed and kubeconfig configured to point to this cluster
```

File: tests/components/istio/README.md [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+- Kyma installed with at least Istio component
+- Export environment variables, the only required is `KYMA_PROFILE`
+
+## Environment variables
```
[the remark this answers is not in the document]
Comment (klaudiagrz) [reply tests/components/istio/README.md:13]: Is that section a part of Prerequisites? If possible, try to structure your README according to the [template](https://github.com/kyma-project/template-repository/blob/main/README.md).

File: tests/components/istio/README.md [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+These environment variables will determine how the tests are run in both prow and your local machine.
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:15]:
```suggestion
These environment variables determine how the tests are run on both Prow and your local machine:
```

File: tests/components/istio/README.md [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+- `KYMA_PROFILE`: Set this environment variable accordingly to the Kyma profile installed on the Kubernetes cluster. These values are: `evaluation` or `production`.
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:17]:
```suggestion
- `KYMA_PROFILE` - set this environment variable accordingly to the Kyma profile installed on the Kubernetes cluster. The possible values are `evaluation` or `production`.
```

File: tests/components/istio/README.md [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+- `EXPORT_RESULT`: Set this environment variable to `true` if you want to export test results to JUnit XML, Cucumber JSON and HTML report. Default value: `false`.
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:18]:
```suggestion
- `EXPORT_RESULT` - set this environment variable to `true` if you want to export test results to JUnit XML, Cucumber JSON, and HTML report. The default value is `false`.
```

File: tests/components/istio/README.md [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+## Executing tests
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:20]:
```suggestion
## Usage
```

File: tests/components/istio/README.md [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+### Start the test suite
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:22]:
```suggestion
1. Start the test suite:
```

File: tests/components/istio/README.md [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+Having prepared environment variables simply run:
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:24]:
```suggestion

```

Comment (klaudiagrz) [reply tests/components/istio/README.md:24]: We already mentioned it in Prerequisites.

File: tests/components/istio/README.md [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+
+`make test`
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:26]:
```suggestion
```make test```
```

File: tests/components/istio/README.md [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+
+We have also provided a Make target that will create k3d cluster, install Kyma on it and run the tests:
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:28]: How is this step related to the next ones?

File: tests/components/istio/README.md [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+
+`make test-k3d`
+
+## Create k3d cluster
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:32]:
```suggestion
2. Provision a k3d cluster on your local machine:
```

File: tests/components/istio/README.md [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+
+To provision k3d cluster on your local machine run:
+
+`make provision-k3d`
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:36]:
```suggestion
```make provision-k3d```
```

File: tests/components/istio/README.md [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+
+## Install Kyma
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:38]:
```suggestion

```

File: tests/components/istio/README.md [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+
+To install Kyma with Istio component on your cluster run:
+
+`make kyma-istio-deploy`
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:42]:
```suggestion
```make kyma-istio-deploy```
```

File: tests/components/istio/main_test.go (hunk 7 above)
Comment (cnvergence) [reply tests/components/istio/main_test.go:63]: we could use it

**Review by barchw — APPROVED** [review]

**Review by barchw — DISMISSED** [review]

File: tests/components/istio/README.md [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+### Environment variables
+
+These environment variables determine how the tests are run on both Prow and your local machine:
+
+- `KYMA_PROFILE` - set this environment variable accordingly to the Kyma profile installed on the Kubernetes cluster. The possible values are `evaluation` or `production`.
+- `EXPORT_RESULT` - set this environment variable to `true` if you want to export test results to JUnit XML, Cucumber JSON, and HTML report. The default value is `false`.
+
+## Usage
+
+### Start the test suite
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:20]:
```suggestion

```

File: tests/components/istio/README.md [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+
+Having met all the requirements simply run:
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:22]:
```suggestion
To start the test suite, run:
```

File: tests/components/istio/README.md [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+
+```make test```
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:24]: Suggestion doesn't work in this case, but could you format this command in such a way?

```
make test
```

To do so, ` ``` ` must be in separate lines.

File: tests/components/istio/README.md [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+
+#### Prepare cluster on your local machine
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:26]: Do we need this section at all? Is it in the scope of the `Istio component tests` document? 🤔

Comment (klaudiagrz) [inline tests/components/istio/README.md:26]:
```suggestion

```

File: tests/components/istio/README.md [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+
+We have also provided a Make target that will create k3d cluster, install Kyma on it and run the tests:
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:28]:
```suggestion
If you don't have a cluster, you can run the tests on your local machine. To do so, run:
```

File: tests/components/istio/README.md [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+
+```make test-k3d```
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:30]: Same as above, please correct the formatting to:

```
make test-k3d
```

File: tests/components/istio/README.md [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+
+The steps are as follows:
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:32]:
```suggestion
This command creates a k3d cluster on your local machine, installs Kyma on it, and runs the tests.
```

File: tests/components/istio/README.md [hunk 29]
```diff
@@ -0,0 +1,44 @@
...
+
+```make provision-k3d```
+
+2. Install Kyma with the Istio component on your cluster:
+
+```make kyma-istio-deploy```
+
+3. Run the tests:
+
+```make test```
```
Comment (klaudiagrz) [inline tests/components/istio/README.md:44]: Remove

**Review by klaudiagrz — APPROVED** [review]

**Review by veichtj — APPROVED** [review]
